helm-v3/006-repositories.md

```diff
@@ -1,40 +1,178 @@
-similar manner to `helm fetch`. It will support different systems via the
-URI/URL scheme and provide a default implementation for HTTP(S). Other systems,
-such as S3, can implement an uploader via a plugin for schemes Helm does not
-provide an implementation for.
-
-> [name=Matt Farina] There very well may be additional work here around
-> registry search and pluggable or otherwise extendable authenication
-> mechanisms.
-> 
-> [name=Adnan Abdulhussein] What would the default HTTP(S) implementation be?
-> [name=Matt Fisher] Adnan, would you mind clarifying your comment a little?
-> 
-> [name=Adnan Abdulhussein] Sorry, the proposal mentions there will be a
-> default implemention for `helm push` to a regular HTTP(S) chart repository.
-> It's unclear to me how this would actually work.
-> 
-> [name=Adnan Abdulhussein] Sounds like this is something that will be fleshed
-> out more in Josh's proposal, so I'll wait for that :)
+- A repository v2 specification will be defined, and the index file will be
+JSON (`index.json` vs `index.yaml`)
+
+## Capabilities API
+
+The Capabilties API is designed to provide Helm with information such as:
+
+1. Does this domain support actions such as login or push?
+2. What supported method should the client use to perform a specific action?
+3. What metadata does the client need to initiate the action?
+
+This will allow Helm to support any number of things in the future related
+to repositories without making breaking changes (e.g. remote `helm search`).
+
+### `/.well-known/helm`
+
+The Capabilities API will be enabled for a specific domain via
+[Well-Known URI](https://tools.ietf.org/html/rfc5785) located at
+`/.well-known/helm`.
+
+*Note: This file may also serve to provide configuration for other APIs
+in the future by supplying an additional top-level section.*
+
+The response body should be valid JSON and should supply a valid,
+top-level `capabilities` section. For example:
+
+```
+{
+    "capabilities": {
+        "login": {
+            ...
+        },
+        "push": {
+            ...
+        },
+        "search": {
+            ...
+        }
+    }
+}
+```
+
+Each subsection (i.e. `login`, `push`, `search`) will provide configuration
+specific to the capabiltity.
+
+For example, the following might inform Helm that this domain only
+accepts token-based login:
+
+```
+...
+        "login": {
+            "methods": {
+                "token": {
+                    "loginUrl": "https://auth.site.com/oauth2/authorize",
+                    "tokenUrl": "https://auth.site.com/oauth2/token"
+                }
+            }
+        }
+...
+```
+
+## The `helm login` command
+
+A new `helm login` command will be added to Helm, responsible for
+managing authentication across repository providers.
+
+The basic usage will be:
+
+```
+helm login <url>
+```
+
+This command is intended to allow users to authenticate just one time in
+order to access multiple repositories provided by the same source. 
```

Comment:
wait, I thought `helm repo add` was being removed in favour of `helm login`. As in, `helm login` is intended as the authentication endpoint for a particular "helm registry", but from that point on no longer needs to `helm repo add`; they just have access to anything that user's auth token is granted to pull/push/view. My vision was similar to how one logs into a docker registry:

```
docker login quay.io
docker pull quay.io/bacongobbler/helloworld
```

Translating to helm, that would look something like this:

```
helm login quay.io
helm install --name helloworld quay.io/bacongobbler/helloworld
```

Is there a particular technical limitation that we need to address via `helm repo add` which blocks this from happening, or have there been other discussions around `helm repo add` where I'm missing this context?

Reply:
I think it's all up for debate. Are you implying that the entire `helm repo` subcommand would/should be removed in Helm 3? 

What effect will this have on support for static repos?

Are we telling users that the only way to install old charts is to do
```
helm install https://kubernetes-charts.storage.googleapis.com/zetcd-0.1.0.tgz
```
vs.
```
helm repo add stable https://kubernetes-charts.storage.googleapis.com/
helm install stable/zetcd
```
?

Reply:
Yeah I was thinking the entire `helm repo` subcommands would be removed in Helm 3 in favour of `helm login`, `helm logout` and `helm search`.

As far as affecting static repos, that's where the capabilities API could come in handy :)

In git, they also have something called "capabilities". If the capabilities API is present, then they assume the "smart" protocol (that is, authentication, streaming APIs, `git push`, the whole kit and kaboodle). However, if that capabilities API is *not* present (such as a git repo hosted on apache/nginx), then they assume the extreme base case (`git pull`, but not `git push`). With Helm, if a status code outside of 200-399 is returned, then we could assume that the server has no capabilities and should only work for `helm fetch`/`helm install`.

Here's a few example CLI workflows:

1. Smart helm repo (chartmuseum)

```
$ helm login https://mychartmuseum.example.com
> GET /.well-known/helm
< 200 OK
{
    "capabilities": {
        "auth": {
            "mechanisms": {
                "2fa": {...}
            }
        }
    }
}

*perform 2-factor auth protocol*
...
OK
$ helm fetch mychartmuseum.example.com/bacongobbler/foobar
```

2. Static server

```
$ helm login https://staticserver.example.com
> GET /.well-known/helm
< 404 NOT FOUND

Warning: server responded with no capabilities. Assuming static server
$ helm fetch staticserver.example.com/bacongobbler/foobar
OK
$ helm push staticserver.example.com/bacongobbler/foobar
Error: server does not support capability "helm-push"
```

Does that make sense?

Reply:
I think the alias is possibly a good balance. The problem I get into today is the local cache has be updated. If I push a new chart, I have to do `helm repo update`.
The term repo is also something I'm hoping we can change. Where a repo is a single chart. A registry is a collection of charts. A repo can only have different versions. hello-world:1, 
So, perhaps
`helm alias add stable kubernetes-charts.storage.googleapis.com/`
`helm install stable/zetcd`
Similar to how docker parses login urls, if the element to the left of the first / doesn't container ., it's assumed to be an alias
FWIW, I'd suggest we don't need http:// as a preface. I would hope that any registry would support SSL. 
Oh, and helm install would use the tag reference for versions:
`helm install kubernetes-charts.storage.googleapis.com/zetcd:0.1.0`

Reply:
Spitballing here: If we match the "dumb" protocol with the legacy API, we might be able to support Helm 2 repositories in Helm 3. I'll do some more digging and try to come up with an answer for you.

Reply:
The biggest opportunity we may be able to do is make Helm 2 forward compatible. Can a helm 2 client send enough info for a helm 3 repo to know how to reply?

Reply:
If someone was willing to put in the work I don't see a technical reason why Helm 2 couldn't support it. I don't think there has been any interest from the core maintainers or the community on supporting helm 3's proposed spec for helm 2, but if someone was interested in picking that up that might be a nice enhancement.

My only concern would be that there was a tentative plan to drop Helm 2 support once Helm 3 is considered stable and enough time was given to users to upgrade. Back-porting features like this may end up forcing the maintainers to support Helm 2 for much longer than anticipated.

Reply:
If there are a large number of charts in a repository being produced by a CI system. I would recommend a way for helm CLI to directly reference a chart without having any local cache or index.json. 
Is there a server side API spec which allows a client to pull a specific chart without downloading any kind of local index?